presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateSerializer.java

```diff
@@ -0,0 +1,282 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AccumulatorStateSerializer;
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.security.InvalidParameterException;
+
+import static com.facebook.presto.spi.StandardErrorCode.CONSTRAINT_VIOLATION;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static java.lang.String.format;
+
+/**
+ * Serializes and creates differential-entropy states.
+ */
+public class StateSerializer
+        implements AccumulatorStateSerializer<State>
+{
+    public static StateStrategy create(
+            long size,
+            double sample)
+    {
+        return new UnweightedReservoirSampleStateStrategy(size);
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight)
+    {
+        return new WeightedReservoirSampleStateStrategy(size);
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight,
+            String method)
+    {
+        if (method.equalsIgnoreCase("reservoir_vasicek")) {
+            return new WeightedReservoirSampleStateStrategy(size);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight,
+            String method,
+            double min,
+            double max)
+    {
+        if (method.equalsIgnoreCase("fixed_histogram_mle")) {
+            return new FixedHistogramMLEStateStrategy(size, min, max);
+        }
+
+        if (method.equalsIgnoreCase("fixed_histogram_jacknife")) {
+            return new FixedHistogramJacknifeStateStrategy(size, min, max);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static void validate(StateStrategy strategy, String method)
+    {
+        if (method.equalsIgnoreCase("reservoir_vasicek")) {
+            if (strategy instanceof UnweightedReservoirSampleStateStrategy) {
+                return;
+            }
+
+            if (strategy instanceof WeightedReservoirSampleStateStrategy) {
+                return;
+            }
+
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    "Inconsistent method");
```

Comment:
This error message is not very descriptive. User wouldn't know what to do with it. Is this actually a user problem or a bug. If this is a bug, then use verify.

`verify(false, "Strategy class " + strategy.getClass().getSimpleName() + " is not compatible with entropy method " + method);`